aiogram/dispatcher/event/handler.py

```diff
@@ -1,45 +1,93 @@
 import asyncio
 import contextvars
 import inspect
+from contextlib import AsyncExitStack
 from dataclasses import dataclass, field
 from functools import partial
-from typing import Any, Awaitable, Callable, Dict, List, Optional, Tuple, Type, Union
+from typing import Any, Awaitable, Callable, Dict, List, Optional, Tuple, Type, Union, cast
 
 from aiogram.dispatcher.filters.base import BaseFilter
 from aiogram.dispatcher.handler.base import BaseHandler
+from aiogram.dispatcher.requirement import (
+    CacheKeyType,
+    Requirement,
+    get_reqs_from_callable,
+    get_reqs_from_class,
+)
 
 CallbackType = Callable[..., Awaitable[Any]]
 SyncFilter = Callable[..., Any]
 AsyncFilter = Callable[..., Awaitable[Any]]
 FilterType = Union[SyncFilter, AsyncFilter, BaseFilter]
 HandlerType = Union[FilterType, Type[BaseHandler]]
 
+REQUIREMENT_CACHE_KEY = "_req_cache"
+ASYNC_STACK_KEY = "_stack"
+
+
+def _is_class_handler(handler: HandlerType) -> bool:
+    return isinstance(handler, type) and issubclass(handler, BaseHandler)
+
 
 @dataclass
 class CallableMixin:
     callback: HandlerType
     awaitable: bool = field(init=False)
     spec: inspect.FullArgSpec = field(init=False)
 
+    __reqs__: Dict[str, Requirement[Any]] = field(init=False)
+
     def __post_init__(self) -> None:
         callback = inspect.unwrap(self.callback)
         self.awaitable = inspect.isawaitable(callback) or inspect.iscoroutinefunction(callback)
         if isinstance(callback, BaseFilter):
-            # Pydantic 1.5 has incorrect signature generator
-            # Issue: https://github.com/samuelcolvin/pydantic/issues/1419
-            # Fixes: https://github.com/samuelcolvin/pydantic/pull/1427
-            # TODO: Remove this temporary fix
-            callback = inspect.unwrap(callback.__call__)
+            callback = callback.__call__
```

Comment:
Here `inspect.unwrap` should be used because callback can be wrapped by decorators and can modify method specs and removing `unwrap` can generate unexpected behavior for class based filters with decorated `__call__` method

Reply:
seems like can be resolved as of pydantic 1.6.x